crates/turborepo-lib/src/engine/builder.rs

```diff
@@ -0,0 +1,985 @@
+            HashSet::from_iter(self.tasks.iter());
+        let mut traversal_queue = VecDeque::with_capacity(1);
+        for (workspace, task) in self.workspaces.iter().cartesian_product(self.tasks.iter()) {
+            let task_id = task
+                .task_id()
+                .unwrap_or_else(|| TaskId::new(workspace.as_ref(), task.task()));
+
+            if self.has_task_definition(&mut turbo_jsons, workspace, task, &task_id)? {
+                missing_tasks.remove(task);
+
+                // Even if a task definition was found, we _only_ want to add it as an entry
+                // point to the task graph (i.e. the traversalQueue), if
+                // it's:
+                // - A task from the non-root workspace (i.e. tasks from every other workspace)
+                // - A task that we *know* is rootEnabled task (in which case, the root
+                //   workspace is acceptable)
+                if !matches!(workspace, WorkspaceName::Root)
+                    || self.root_enabled_tasks.contains(task)
+                {
+                    traversal_queue.push_back(task_id);
+                }
+            }
+        }
+
+        if !missing_tasks.is_empty() {
+            return Err(Error::MissingTasks(missing_tasks.into_iter().join(", ")));
+        }
+
+        let mut visited = HashSet::new();
+        let mut engine = Engine::default();
+
+        while let Some(task_id) = traversal_queue.pop_front() {
+            if task_id.package() == ROOT_PKG_NAME
+                && !self
+                    .root_enabled_tasks
+                    .contains(&task_id.as_non_workspace_task_name())
+            {
+                return Err(Error::MissingTaskForRoot {
+                    task_id: task_id.to_string(),
+                });
+            }
+
+            if task_id.package() != ROOT_PKG_NAME
+                && self
+                    .package_graph
+                    .package_json(&WorkspaceName::from(task_id.package()))
+                    .is_none()
+            {
+                // If we have a pkg it should be in WorkspaceInfos.
+                // If we're hitting this error something has gone wrong earlier when building
+                // WorkspaceInfos or the workspace really doesn't exist and
+                // turbo.json is misconfigured.
+                return Err(Error::MissingWorkspaceFromTask {
+                    package: task_id.package().to_string(),
+                    task_id: task_id.to_string(),
+                });
+            }
+            let task_definition = Rc::new(TaskDefinition::from_iter(self.task_definition_chain(
+                &mut turbo_jsons,
+                &task_id,
+                &task_id.as_non_workspace_task_name(),
+            )?));
+
+            // Skip this iteration of the loop if we've already seen this taskID
+            if visited.contains(&task_id) {
```

Comment:
This check happens here in the Go code. I believe it can happen earlier, but I didn't want to shuffle around control flow during the port.